src/jpeg2000Compression.cpp

```diff
@@ -28,21 +30,49 @@ std::string Jpeg2000Compression::toString() const {
   return std::string("lossless JPEG2000 compressed");
 }
 
+
+static void openjpeg_error(const char* msg, void* client_data) {
+  BOOST_LOG_TRIVIAL(error) << "JPEG 2000 Error: " << msg;
+}
+
+static void openjpeg_warning(const char* msg, void* client_data) {
+  BOOST_LOG_TRIVIAL(warning) << "JPEG 2000 Warning: " << msg;
+}
+
+static void openjpeg_info(const char* msg, void* client_data) {
+  BOOST_LOG_TRIVIAL(info) << "JPEG 2000 Info: " << msg;
+}
+
 std::unique_ptr<uint8_t[]> Jpeg2000Compression::writeToMemory(
-    unsigned int width, unsigned int height, unsigned int pitch,
```

Comment:
Yes minor tangential change.  The pixel memory (bytes) hits this function.  With a layout ~ RGBRGBRGBRGB
Layed out in 2D it would be, RGBRGBRGBRGB
                                                RGBRGBRGBRGB
                                                .....
                                                                 .... RGB

Total bytes in the first row = 3 (RGB) * pixel width.
Total bytes in the image = 3 * width * height.

The JPEG2000 codec, at least as it is used here, consumes the R, G, and B as independent arrays.
The RGB triples correspond by index only.

plane 1 (red) = RRRRRRRRRRRRRRR   
plane 2 (green) = GGGGGGGGGGG   
plane 3 (blue) = BBBBBBBBBBBBB

This as the old, loop in the code which converted the input data into the format that JPEG2000 wanted.
Pitch = 3 * width, which was just the number of bytes in a row of the memory.  The loop then computed the byte offset into the interlaced (RGBRGB) memory.  base address + y * pitch, and then walked across the row.  All, I did here was simplify it, and walk over all the bytes in one loop removing the pitch, multiplication.


 int32_t* red = opjImage->comps[0].data;
 int32_t* green = opjImage->comps[1].data;
 int32_t* blue = opjImage->comps[2].data;
 for (unsigned int y = 0; y < height; y++) {
    const uint8_t* pixel = reinterpret_cast<const uint8_t*>(buffer) + y * pitch;

    for (unsigned int x = 0; x < width; x++) {
      *red = pixel[0];
      *green = pixel[1];
      *blue = pixel[2];
      pixel += 3;
      red++;             
      green++;
      blue++;
    }

red, green and blue are initialized to the allocated base address where the per channel memory will be copied (I assume this what it is doing, I didn't write it myself).
*red  = pixel[0], copies the value in in the pixel arrays first byte into the memory pointed to by the pointer.  
later, the call to red++, increments the pointer, by the size of the type it points to.  This will then result in the next value being copied into the pointers index 1 position.  All of this is safe only as along as the buffers pointed to by red, blue, and green are preallocated to be the  sizeof (type pointed to) * width * height.

Reply:
All of this is is a nice example of how C/C++ can be a very fast language and also a dangerous one.  At best we would segfault if we accessed memory outside of our allocation and at worst cause a bug which would be difficult to debug.

Reply:
I see, thanks for the explanation!